src/OrchardCore/OrchardCore/Modules/ModularTenantRouterMiddleware.cs

```diff
@@ -49,37 +51,33 @@ public async Task Invoke(HttpContext httpContext)
                 httpContext.Request.Path = httpContext.Request.Path.ToString().Substring(httpContext.Request.PathBase.Value.Length);
             }
 
-            // TODO: Invalidate the pipeline automatically when the shell context is changed
-            // such that we can reload the middlewares and the routes. Implement something similar
-            // to IRunningShellTable but for the pipelines.
-
             // Do we need to rebuild the pipeline ?
-            var rebuildPipeline = httpContext.Items["BuildPipeline"] != null;
-            if (rebuildPipeline && _pipelines.ContainsKey(shellContext.Settings.Name))
+            if (!shellContext.IsPipelineBuilt)
```

Comment:
Should the pipeline be on the shell context? And this way it would also be disposed when reloaded ?